zio-schema/shared/src/main/scala/zio/schema/DynamicValue.scala

```diff
@@ -2087,203 +2087,260 @@ private[schema] object DynamicValueSchema { self =>
+      }
     )
 
   private val primitiveIntCase: Schema.Case[DynamicValue.Primitive[Int], DynamicValue] =
     Schema.Case(
       "Int",
-      Schema.primitive[Int].transform(i => DynamicValue.Primitive(i, StandardType[Int]), _.value),
-      _.asInstanceOf[DynamicValue.Primitive[Int]]
+      Schema.primitive[Int].transform(i => DynamicValue.Primitive(i, StandardType[Int]), _.value), {
+        case dv @ DynamicValue.Primitive(_: Int, _) => dv.asInstanceOf[DynamicValue.Primitive[Int]]
+        case _                                      => throw new IllegalArgumentException
+      }
     )
 
   private val primitiveLongCase: Schema.Case[DynamicValue.Primitive[Long], DynamicValue] =
     Schema.Case(
       "Long",
-      Schema.primitive[Long].transform(l => DynamicValue.Primitive(l, StandardType[Long]), _.value),
-      _.asInstanceOf[DynamicValue.Primitive[Long]]
+      Schema.primitive[Long].transform(l => DynamicValue.Primitive(l, StandardType[Long]), _.value), {
+        case dv @ DynamicValue.Primitive(_: Long, _) => dv.asInstanceOf[DynamicValue.Primitive[Long]]
+        case _                                       => throw new IllegalArgumentException
+      }
     )
 
   private val primitiveFloatCase: Schema.Case[DynamicValue.Primitive[Float], DynamicValue] =
     Schema.Case(
       "Float",
-      Schema.primitive[Float].transform(f => DynamicValue.Primitive(f, StandardType[Float]), _.value),
-      _.asInstanceOf[DynamicValue.Primitive[Float]]
+      Schema.primitive[Float].transform(f => DynamicValue.Primitive(f, StandardType[Float]), _.value), {
+        case dv @ DynamicValue.Primitive(_: Float, _) => dv.asInstanceOf[DynamicValue.Primitive[Float]]
+        case _                                        => throw new IllegalArgumentException
+      }
     )
 
   private val primitiveDoubleCase: Schema.Case[DynamicValue.Primitive[Double], DynamicValue] =
     Schema.Case(
       "Double",
-      Schema.primitive[Double].transform(d => DynamicValue.Primitive(d, StandardType[Double]), _.value),
-      _.asInstanceOf[DynamicValue.Primitive[Double]]
+      Schema.primitive[Double].transform(d => DynamicValue.Primitive(d, StandardType[Double]), _.value), {
+        case dv @ DynamicValue.Primitive(_: Double, _) => dv.asInstanceOf[DynamicValue.Primitive[Double]]
+        case _                                         => throw new IllegalArgumentException
+      }
     )
 
   private val primitiveBinaryCase: Schema.Case[DynamicValue.Primitive[Chunk[Byte]], DynamicValue] =
     Schema.Case(
       "Binary",
-      Schema.primitive[Chunk[Byte]].transform(ch => DynamicValue.Primitive(ch, StandardType[Chunk[Byte]]), _.value),
-      _.asInstanceOf[DynamicValue.Primitive[Chunk[Byte]]]
+      Schema.primitive[Chunk[Byte]].transform(ch => DynamicValue.Primitive(ch, StandardType[Chunk[Byte]]), _.value), {
+        case dv @ DynamicValue.Primitive(_: Chunk[_], _) => dv.asInstanceOf[DynamicValue.Primitive[Chunk[Byte]]]
```

Comment:
I think it's safe to treat Chunk[Byte] as Chunk[_] since the information about Chunk[Byte] is going to be erased anyway

Reply:
@thinkharderdev Should I squash all the commits? Unfortunately, because of some renaming in the `StandardType` I got a lot of changes in all those places where StandardType values are used